### PR TITLE
[AzSdkCli] Add client.tsp fixes loop to `typespec client customized-update` tool

### PR DESCRIPTION
**Pull request by m-redding**
## Summary

Adds an AI-assisted `client.tsp` customization loop to the `typespec client customized-update` tool. When a user requests TypeSpec customizations, the tool now classifies changes, applies `client.tsp` modifications, regenerates the SDK, builds, and retries — before falling back to SDK-level patching.

### Core: `CustomizedCodeUpdateTool`
- **TypeSpec customization loop**: Classify feedback → apply `client.tsp` changes via `ITypeSpecCustomizationService` → regenerate → build → retry (up to 2 iterations) with build error context appended
- **Classification-driven routing**: Items are categorized as `TSP_APPLICABLE`, `REQUIRES_MANUAL_INTERVENTION`, or `SUCCESS`. Only TSP-applicable items enter the customization loop; manual items are reported with next-step guidance
- **Input validation**: `tspProjectPath` validated via `typeSpecHelper.IsValidTypeSpecProjectPath()` (checks directory + tspconfig.yaml). `customizationRequest` is now required (non-nullable)
- **Error codes**: Added `ManualInterventionRequired`, `TypeSpecCustomizationFailed` to `KnownErrorCodes` for machine-readable downstream processing

### Helpers
- `ITspClientHelper.UpdateGenerationAsync` / `TspClientHelper`: Added `localSpecRepoPath` parameter, forwarded as `--local-spec-repo` to tsp-client
- `ITypeSpecCustomizationService` + `ITypeSpecHelper` injected into tool

### Agent Runner (`CopilotAgentRunner`)
- Added `OnPermissionRequest` handler to auto-approve tool execution — required for autonomous agent operation (SDK returns "Permission denied" without it)
- Bumped `GitHub.Copilot.SDK` to `0.1.29`

### Response Model (`CustomizedCodeUpdateResponse`)
- Added `TypeSpecChangesSummary` to surface `client.tsp` edits made during the loop
- Added new error codes for finer-grained failure reporting

### Tests
- 21 tests covering the full loop: happy path, retry, classification edge cases, validation, partial failures, manual intervention, regeneration failures
- Updated `SdkGenerationToolTests` and `CopilotAgentRunnerTests` for new signatures

### Other
- Updated evaluation prompts in `TestPrompts.json` to match new tool intent
- `FeedbackClassificationResponse.ItemClassificationDetails.Text` made non-nullable

### CODE REVIEW

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 1]
```diff
@@ -114,6 +139,66 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+                    return new CustomizedCodeUpdateResponse
+                    {
+                        Message = $"TypeSpec customization failed: {tspCustomizationResult.FailureReason}",
+                        ErrorCode = CustomizedCodeUpdateResponse.KnownErrorCodes.TypeSpecCustomizationFailed,
+                        ResponseError = tspCustomizationResult.FailureReason
+                    };
+                }
+
+                logger.LogInformation("TypeSpec customization succeeded. Changes: {Changes}", string.Join("; ", tspCustomizationResult.ChangesSummary));
+                var typeSpecChanges = tspCustomizationResult.ChangesSummary.ToList();
```
Comment (swathipil) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: I think we'll also want to parse the changes so that each line in the summary can be mapped to its corresponding feedback via the "id". Then we can add that to the feedback context using this method: https://github.com/Azure/azure-sdk-tools/blob/82642a7b7e4cc28c204eaa1115b2faf70fa52b51/tools/azsdk-cli/Azure.Sdk.Tools.Cli/Models/FeedbackData.cs#L63

The classifier will use this context on the next iteration of the loop to reclassify the items.

Comment (swathipil) [reply tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: e.g.
```
------ITERATION 1-----
=== APPLY TYPESPEC CUSTOMIZATIONS ===
Renamed X to Y.
=== End APPLY TYPESPEC CUSTOMIZATIONS ===
=== GENERATE SDK ===
Python emitter successfully generated. <insert other emitter output>
=== End GENERATE SDK ===
=== BUILD SDK ===
pylint errors: 15...
=== End BUILD SDK ===
----ITERATION 2----
```
which you could add by calling the following in the appropriate locations:
```csharp
feedback.AppendContext("------ITERATION 1-----")
feedback.AppendContext("Renamed X to Y", "TYPESPEC CUSTOMIZATIONS")
feedback.AppendContext(emitterOutput, "GENERATE SDK")
feedback.AppendContext(buildOutput, "BUILD SDK")
```

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Helpers/TspClientHelper.cs [hunk 2]
```diff
@@ -79,28 +79,34 @@ public async Task<TspToolResponse> ConvertSwaggerAsync(string swaggerReadmePath,
     }
 
     /// <inheritdoc />
-    public async Task<TspToolResponse> UpdateGenerationAsync(string tspLocationDirectory, string? commitSha = null, bool isCli = false, CancellationToken ct = default)
+    public async Task<TspToolResponse> UpdateGenerationAsync(string tspLocationDirectory, string? commitSha = null, bool isCli = false, string? localSpecRepoPath = null, CancellationToken ct = default)
     {
         var tspLocationPath = Path.Combine(tspLocationDirectory, "tsp-location.yaml");
-        logger.LogInformation("tsp-client update: {tspLocationDirectory}, commit: {commit}", tspLocationDirectory, commitSha ?? "(latest)");
-        
+        logger.LogInformation("tsp-client update: {tspLocationDirectory}, commit: {commit}, localSpecRepo: {localSpecRepo}", tspLocationDirectory, commitSha ?? "(latest)", localSpecRepoPath ?? "(none)");
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Helpers/TspClientHelper.cs]: nit: I think this doubly logs. The command runner does log the command that its about to run.

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Helpers/TspClientHelper.cs [hunk 3]
```diff
@@ -79,28 +79,34 @@ public async Task<TspToolResponse> ConvertSwaggerAsync(string swaggerReadmePath,
     }
 
     /// <inheritdoc />
-    public async Task<TspToolResponse> UpdateGenerationAsync(string tspLocationDirectory, string? commitSha = null, bool isCli = false, CancellationToken ct = default)
+    public async Task<TspToolResponse> UpdateGenerationAsync(string tspLocationDirectory, string? commitSha = null, bool isCli = false, string? localSpecRepoPath = null, CancellationToken ct = default)
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Helpers/TspClientHelper.cs:82]: what happens when localSpecRepoPath not provided?

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 4]
```diff
@@ -44,31 +48,52 @@ public CustomizedCodeUpdateTool(
...
         _classifierService = classifierService;
+        this.typeSpecCustomizationService = typeSpecCustomizationService;
+        this.typeSpecHelper = typeSpecHelper;
     }
 
     public override CommandGroup[] CommandHierarchy { get; set; } = [SharedCommandGroups.TypeSpec, SharedCommandGroups.TypeSpecClient];
 
+    private readonly Option<string?> customizationRequestOption = new("--customization-request")
+    {
+        Description = "Optional description of requested customization to apply to the TypeSpec."
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: @swathipil Imo this is not optional, could you confirm?

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 5]
```diff
@@ -44,31 +48,52 @@ public CustomizedCodeUpdateTool(
...
+    };
+
+    private readonly Option<string?> typespecProjectPath = new("--tsp-project-path")
+    {
+        Description = "Optional description of requested customization to apply to the TypeSpec."
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: not the right description?

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 6]
```diff
@@ -271,4 +343,60 @@ private async Task<FeedbackClassificationResponse> Classify(
...
+        {
+            return (false, "No feedback was found to apply.");
+        }
+
+        StringBuilder tspApplicableFeedback = new();
+        foreach (var feedback in response.Classifications)
+        {
+            if (feedback.Classification == "TSP_APPLICABLE")
+            {
+                if (feedback.Text == null)
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: @swathipil we should fix this. If we have a classification the feedback text cannot be empty.

**Review by samvaity — COMMENTED** [review]
changes mostly lgtm

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 7]
```diff
@@ -271,4 +344,57 @@ private async Task<FeedbackClassificationResponse> Classify(
...
+                var tspCustomizationResult = await typeSpecCustomizationService.ApplyCustomizationAsync(tspProjectPath, feedback.Text.ToString(), ct: ct);
+
+                if (!tspCustomizationResult.Success)
+                {
+                    logger.LogWarning("TypeSpec customization failed: {Reason}", tspCustomizationResult.FailureReason);
+                    return (false, tspCustomizationResult.FailureReason);
+                }
+
+                logger.LogInformation("TypeSpec customization succeeded. Changes: {Changes}", string.Join("; ", tspCustomizationResult.ChangesSummary));
+                var typeSpecChanges = tspCustomizationResult.ChangesSummary.ToList();
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: where do we use this?

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs (hunk 1 above)
Comment (m-redding) [reply tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: Ok I updated the plain text input to give to the classifier to inlcude information about updated changes, failures, build failures, etc. We can try this for now but this will need to be reworked once we start working with API view changes!

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs (hunk 4 above)
Comment (m-redding) [reply tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: Fixed

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Helpers/TspClientHelper.cs (hunk 3 above)
Comment (m-redding) [reply tools/azsdk-cli/Azure.Sdk.Tools.Cli/Helpers/TspClientHelper.cs:82]: --local-spec-repo arg is omitted from the tsp-client update call. The tool then falls back to fetching the spec remotely using the repo/commit info from tsp-location.yaml.

I added a comment to ITspClientHelper

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 8]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+        var feedbackDictionary = feedbackItems.ToDictionary(i => i.Id, i => i);
+
+        List<string> changesMade = new();
+        StringBuilder noActionNeeded = new();
+        noActionNeeded.AppendLine("The following feedback items were classified as SUCCESS (no changes needed):");
+        List<string> manualInterventions = new();
+        StringBuilder tspFixFailedReasons = new();
+        do
+        {
+            // TODO - need to update this to avoid casting to/from list
```
Comment (m-redding) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs:208]: This is a future todo since it will involve taking a holistic look at the tool now that it's all hooked up together and seeing how we want to pass data through each piece

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 9]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+            var noChanges = 0;
+
+            foreach (var itemDetails in response.Classifications)
+            {
+                feedbackDictionary.TryGetValue(itemDetails.ItemId, out var feedbackItem);
+
+                // For now, if feedback item isn't found we just won't append context
+                feedbackItem?.AppendContext($"Iteration {tries+1}");
+
+                if (itemDetails.Classification == "TSP_APPLICABLE")
```
Comment (m-redding) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: TODO - I should put these constants as consts

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 10]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+                    };
+                }
+            }
+
+            if (tspFixFailed > 0)
+            {
+                logger.LogWarning("Some customizations failed to apply: {FailureReasons}", tspFixFailedReasons);
+            }
+
+            // Don't waste time regenerating if no TSP files were successfully applied
```
Comment (m-redding) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]:
```suggestion
            // Don't waste time regenerating if no TSP fixes were successfully applied
```

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 11]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+            }
+
+            var (success, error, _) = await languageService.BuildAsync(packagePath, CommandTimeoutInMinutes, ct);
+
+            buildSucceeded = success;
+            buildError = error;
+
+            // Append build result context to all remaining feedback items for the next iteration
+            var buildContext = success ? "Build succeeded." : (error ?? "Build failed with unknown error.");
+            feedbackDictionary.Values.ToList().ForEach(item => item.AppendContext(buildContext, "Build Result"));
```
Comment (m-redding) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: TODO - this shouldn't happen on the last try

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 12]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+
+            // Exit cases for the first attempt
+            if (tries == 0)
+            {
+                // Nothing was classified as tsp applicable and at least some feedback requires manual intervention
+                if (tspApplicable == 0 && manualChanges > 0)
+                {
+                    return new CustomizedCodeUpdateResponse
+                    {
+                        Success = false,
```
Comment (m-redding) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: Maybe this should be a success response

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/CopilotAgents/CopilotAgentRunner.cs [hunk 13]
```diff
@@ -92,7 +92,15 @@ public async Task<TResult> RunAsync<TResult>(
             },
             // Disables copilot built-in tools
             // copilot built-in tools may be enabled in the future as needed
-            AvailableTools = [.. tools.Select(t => t.Name)]
+            AvailableTools = [.. tools.Select(t => t.Name)],
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/CopilotAgents/CopilotAgentRunner.cs:95]: Thank you for fixing this @m-redding !

Ref this was the error we were getting
>Agent completed without calling Exit tool (attempt 2/3). Prompting to call Exit.
...
Classification failed: Agent failed to call Exit tool after 3 reminders

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 14]
```diff
@@ -18,12 +21,14 @@ namespace Azure.Sdk.Tools.Cli.Tools.TypeSpec;
 /// MCP tool that updates SDK code from TypeSpec, applies patches to customization files,
 /// regenerates code, builds, and provides intelligent analysis and recommendations for updating customization code.
 /// </summary>
-[McpServerToolType, Description("Updates SDK code from TypeSpec, applies patches to customization files, regenerates code, builds, provides intelligent analysis and recommendations for updating customization code.")]
+[McpServerToolType, Description("Apply TypeSpec and SDK code customizations: updates client TypeSpec or SDK code, provides code update recommendations, and regenerates SDK packages.")]
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs:23]: Since you updated the description, could you please update the references in TestPrompts.json as well? There are tool match errors right now because I forgot to do this last time.

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 15]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+            // Exit cases for the first attempt
+            if (tries == 0)
+            {
+                // Nothing was classified as tsp applicable and at least some feedback requires manual intervention
+                if (tspApplicable == 0 && manualChanges > 0)
+                {
+                    return new CustomizedCodeUpdateResponse
+                    {
+                        Success = false,
+                        Message = "None of the requested changes could be fixed with TypeSpec customizations",
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: yes, success. And we could also make this mention manual intervention phrases since that is the tooling's resolution?

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 16]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+            {
+                feedbackDictionary.TryGetValue(itemDetails.ItemId, out var feedbackItem);
+
+                // For now, if feedback item isn't found we just won't append context
+                feedbackItem?.AppendContext($"Iteration {tries+1}");
+
+                if (itemDetails.Classification == "TSP_APPLICABLE")
+                {
+                    tspApplicable++;
+                    var tspCustomizationResult = await typeSpecCustomizationService.ApplyCustomizationAsync(tspProjectPath, itemDetails.Text, ct: ct);
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs:250]: Are we applying customizations once for all or one-by-one? In case of each handling do we remove it from the feedback batch so that we don't iterate on it again?

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 17]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+                    };
+                }
+
+                // Everything was classified as success
+                if (tspApplicable == 0 && noChanges > 0)
+                {
+                    return new CustomizedCodeUpdateResponse
+                    {
+                        Success = true,
+                        Message = $"All feedback items were classified as SUCCESS (no changes needed):{Environment.NewLine}{noActionNeeded}"
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: Is this the tooling response or intermediary response?
If tooling response, it should be more general and not mention inner details of classify or feedback items imo. What do you think?

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 18]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+            // Append build result context to all remaining feedback items for the next iteration
+            var buildContext = success ? "Build succeeded." : (error ?? "Build failed with unknown error.");
+            feedbackDictionary.Values.ToList().ForEach(item => item.AppendContext(buildContext, "Build Result"));
+
+            tries++;
+        } while (buildSucceeded == false && tries < maxTries);
+
+        if (buildSucceeded)
         {
+            logger.LogInformation("Build passed - ");
```
Comment (samvaity) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]:
```suggestion
            logger.LogInformation("Build passed after Typespec customizations.");
```

**Review by samvaity — APPROVED** [review]

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs (hunk 17 above)
Comment (m-redding) [reply tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: This is the tooling response, and yes that makes sense

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs (hunk 16 above)
Comment (m-redding) [reply tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs:250]: I was debating on how to do this and am open to feedback on changing this.

Right now, the tool only takes in one piece of feedback so I don't have any data about how it will perform in each respective scenario. To me it seems like the agent might be more successful with a one by one approach, especially looking at how the spec expects the flow to almost be independent for each item. 

IRT removing it - I only remove manual/success. I was following the spec where it talks about appending the build and how the agent tried to fix it the first time - https://github.com/Azure/azure-sdk-tools/blob/main/tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md#context-tracking adding to the context and then passing it back to the classifier to see what should happen next.

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 19]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
             };
         }
 
+        // Get language info
         var languageService = await GetLanguageServiceAsync(packagePath, ct);
-        if (!languageService.IsCustomizedCodeUpdateSupported)
+        // TODO - do this once we add API view option
```
Comment (swathipil) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs:179]: Not sure if this was left commented intentionally?

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 20]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+        // var language = await feedbackService.GetLanguageAsync(apiViewUrl, ct);
+
+        // Step 1: try tsp fixes
+        var tries = 0;
+        var maxTries = 2;
+        bool buildSucceeded = false;
+        string? buildError = null;
+
+        var feedbackItems = await GetFeedbackItems(tspProjectPath, plainTextFeedback: customizationRequest, ct: ct);
```
Comment (swathipil) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: If this is 0, then would we want this to be a "SUCCESS"?

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 21]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+            var tspApplicable = 0;
+            var manualChanges = 0;
+            var noChanges = 0;
+
+            foreach (var itemDetails in response.Classifications)
+            {
+                feedbackDictionary.TryGetValue(itemDetails.ItemId, out var feedbackItem);
+
+                // For now, if feedback item isn't found we just won't append context
+                feedbackItem?.AppendContext($"Iteration {tries+1}");
```
Comment (swathipil) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs:244]: Should we debug log the scenario where it returns a non-existent feedback item ID? That might be helpful in trying to fix hallucincations.

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 22]
```diff
@@ -114,33 +147,199 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+
+            // Don't waste time regenerating if no TSP files were successfully applied
+            if (tspFixSucceeded > 0)
+            {
+                // Regenerate SDK using local spec repo
+                var regenResult = await tspClientHelper.UpdateGenerationAsync(packagePath, localSpecRepoPath: tspProjectPath, isCli: false, ct: ct);
+                if (!regenResult.IsSuccessful)
+                {
+                    logger.LogWarning("Regeneration failed: {Error}", regenResult.ResponseError);
+                    return new CustomizedCodeUpdateResponse
```
Comment (swathipil) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]: If regeneration fails, should it add that to the item's context and `continue` rather than `return` so that the classifier can re-classify it and potentially give steps for manual intervention?

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs (hunk 21 above)
Comment (m-redding) [reply tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs:244]: YEs I should probably add a few more logs in general - will update

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs (hunk 19 above)
Comment (m-redding) [reply tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs:179]: Yes that one I left intentionally since it was part of the original classify method, when I switched to get feedback items I didn't want it to get lost

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs [hunk 23]
```diff
@@ -114,33 +152,200 @@ private async Task<CustomizedCodeUpdateResponse> RunUpdateAsync(string packagePa
...
+            var manualChanges = 0;
+            var noChanges = 0;
+
+            foreach (var itemDetails in response.Classifications)
+            {
+                feedbackDictionary.TryGetValue(itemDetails.ItemId, out var feedbackItem);
+
+                if (feedbackItem == null)
+                {
+                    logger.LogDebug("Classifier returned non-existent feedback item ID '{ItemId}'. This may indicate an LLM hallucination.", itemDetails.ItemId);
```
Comment (m-redding) [inline tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/TypeSpec/CustomizedCodeUpdateTool.cs]:
```suggestion
                    logger.LogDebug("Classifier returned non-existent feedback item ID '{ItemId}'.", itemDetails.ItemId);
```

**Review by m-redding — COMMENTED** [review]
_moved (copilot commented in the wrong place)_

File: tools/azsdk-cli/Azure.Sdk.Tools.Cli/CopilotAgents/CopilotAgentRunner.cs [hunk 24]
```diff
@@ -92,7 +92,15 @@ public async Task<TResult> RunAsync<TResult>(
...
+            // Auto-approve all permission requests since agents run autonomously with no user interaction.
+            // The CLI's permission system sends permission.request RPC calls for tool execution.
+            // Without this handler, the SDK cannot get user consent and returns "Permission denied".
+            OnPermissionRequest = (request, invocation) =>
+            {
+                logger.LogDebug("Auto-approving permission request for tool execution");
+                return Task.FromResult(new PermissionRequestResult { Kind = "approved" });
+            }
```
[the remark this answers is not in the document]
Comment (m-redding) [reply tools/azsdk-cli/Azure.Sdk.Tools.Cli/CopilotAgents/CopilotAgentRunner.cs:103]: Good suggestion for defense-in-depth. I investigated this and here's why we're keeping the blanket approve for now:

AvailableTools (line 95) is the real security boundary — it restricts which tools the model can see and call, and it's already scoped to only our custom MCP tools + the Exit tool. The OnPermissionRequest handler exists solely to unblock the SDK's consent flow, which otherwise returns "Permission denied" for all tool execution (this was the root cause of the "Agent failed to call Exit tool after 3 reminders" bug).

Filtering by tool name in the permission handler has a few issues:

Permission requests have different kind values ("mcp", "shell", "write", "read", "url") — filtering only on tool name could accidentally deny legitimate non-tool requests.
Per https://github.com/github/copilot-sdk/issues/254, AIFunctionFactory tools (like our Exit tool) may bypass the permission handler entirely, so it wouldn't even gate the most critical tool.
Added complexity with limited practical benefit since AvailableTools already prevents the model from calling anything unexpected.
Happy to revisit if the SDK stabilizes its permission model or if we expand the agent's capabilities.

**Review by swathipil — APPROVED** [review]
lgtm!

**Review by samvaity — APPROVED** [review]
